cmd/snap-seccomp-blacklist/snap-seccomp-blacklist.c

```diff
@@ -0,0 +1,215 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <seccomp.h>
+
+__attribute__((format(printf, 1, 2))) static void showerr(const char *fmt, ...);
+
+static void showerr(const char *fmt, ...) {
+    va_list va;
+    va_start(va, fmt);
+    vfprintf(stderr, fmt, va);
+    fputc('\n', stderr);
+    va_end(va);
+}
+
+static int populate_filter(scmp_filter_ctx ctx, const uint32_t *arch_tags, size_t num_arch_tags) {
+    int sc_err;
+
+    /* If the native architecture is not one of the supported 64bit
+     * architectures listed above then remove it.
+     *
+     * Libseccomp automatically adds the native architecture to each new filter.
+     * If the native architecture is a 32bit-one then we will hit a bug in libseccomp
+     * and the generated BPF program is incorrect as described below. */
+    uint32_t native_arch = seccomp_arch_native();
+    bool remove_native_arch = true;
+    for (size_t i = 0; i < num_arch_tags; ++i) {
+        if (arch_tags[i] == native_arch) {
+            remove_native_arch = false;
+            break;
+        }
+    }
+    if (remove_native_arch) {
+        sc_err = seccomp_arch_remove(ctx, SCMP_ARCH_NATIVE);
+        if (sc_err < 0) {
+            showerr("cannot remove native architecture");
+            return sc_err;
+        }
+    }
+
+    /* Add 64-bit architectures supported by snapd into the seccomp filter.
+     *
+     * The documentation of seccomp_arch_add() is confusing. It says that after
+     * this call any new rules will be added to this architecture. This is
+     * correct. It doesn't, however, explain that the rules will be multiplied
+     * and re-written as explained below. */
+    for (size_t i = 0; i < num_arch_tags; ++i) {
+        uint32_t arch_tag = arch_tags[i];
+        sc_err = seccomp_arch_add(ctx, arch_tag);
+        if (sc_err < 0 && sc_err != -EEXIST) {
+            showerr("cannot add architecture %x", arch_tag);
+            return sc_err;
+        }
+    }
+
+    /* When the rule set doesn't match one of the architectures above then the
+     * resulting action should be a "allow" rather than "kill". We don't add
+     * any of the 32bit architectures since there is no need for any extra
+     * filtering there. */
+    sc_err = seccomp_attr_set(ctx, SCMP_FLTATR_ACT_BADARCH, SCMP_ACT_ALLOW);
+    if (sc_err < 0) {
+        showerr("cannot set action for unknown architectures");
+        return sc_err;
+    }
+
+    /* Resolve the name of "ioctl" on this architecture. */
+    int sys_ioctl_nr;
+    sys_ioctl_nr = seccomp_syscall_resolve_name("ioctl");
```

Comment:
I think this could go away if we use SCMP_SYS, see comment below.

Reply:
We cannot, as explained above.